xarray/convert.py

```diff
@@ -157,34 +161,48 @@ def _iris_cell_methods_to_str(cell_methods_obj):
     return ' '.join(cell_methods)
 
 
+def _name(iris_obj, default='unknown'):
+    """ Mimicks `iris_obj.name()` but with different name resolution order.
+
+    Same as iris_obj.name() method, but using iris_obj.var_name first to enable
+    roundtripping.
+    """
+    return iris_obj.var_name or iris_obj.standard_name or \
+        iris_obj.long_name or str(iris_obj.attributes.get('STASH', '')) or \
+        default
+
+
 def from_iris(cube):
     """ Convert a Iris cube into an DataArray
     """
     import iris.exceptions
     from xarray.core.pycompat import dask_array_type
 
-    name = cube.var_name
+    name = _name(cube)
+    if name == 'unknown':
+        name = None
     dims = []
     for i in range(cube.ndim):
         try:
             dim_coord = cube.coord(dim_coords=True, dimensions=(i,))
-            dims.append(dim_coord.var_name)
+            dims.append(_name(dim_coord))
         except iris.exceptions.CoordinateNotFoundError:
             dims.append("dim_{}".format(i))
 
+    if len(set(dims)) != len(dims):
+        from collections import Counter
```

Comment:
Please import this at the top of the module -- we only don't do this for `iris` because Iris is an optional xarray dependency (but `collections` is in the standard library).